File: android/app/src/main/java/com/nickdegs/mobbing/GameEngine.kt
package com.nickdegs.mobbing

import android.content.Context
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlin.math.max
import kotlin.math.min
import kotlin.random.Random

// ---------------------------------------------------------------------------
// Veri modeli — cards_*.json ile aynı şema (DEĞİŞMEDİ, geriye uyumlu)
// ---------------------------------------------------------------------------
@Serializable data class LocText(val en: String, val tr: String? = null) {
    fun get(lang: String): String = if (lang == "tr" && tr != null) tr else en
}
@Serializable data class Choice(val t: LocText, val fx: List<Int>, val next: String? = null)
@Serializable data class Card(
    val id: String, val cat: String, val ch: String,
    val minB: Int? = null, val minD: Int? = null, val t: LocText, val l: Choice, val r: Choice
)
@Serializable data class CardFile(val cards: List<Card>)

// 5 gösterge: baskı, vicdan, ekip, kariyer, SAĞLIK (yeni)
data class Meters(var b: Int = 50, var v: Int = 50, var e: Int = 50, var k: Int = 50, var h: Int = 100)

enum class Ending(val key: String) {
    B0("end_b0"), B100("end_b100"), V0("end_v0"), V100("end_v100"),
    E0("end_e0"), E100("end_e100"), K0("end_k0"), K100("end_k100"),
    H0("end_h0"), LAWSUIT("end_lawsuit"),  // YENİ: sağlık çöküşü + dava zaferi
    CAUGHT("end_caught")                    // YENİ: kanıt toplarken yakalandın (mahvoldun)
}

// ---------------------------------------------------------------------------
// Motor
// ---------------------------------------------------------------------------
class GameEngine(context: Context, private val lang: String) {

    private val allCards: List<Card>
    private val followupIds: Set<String>
    private val queue: ArrayDeque<String> = ArrayDeque()
    private val rng = Random(System.currentTimeMillis())

    val meters = Meters()
    var day = 1; private set
    var current: Card? = null; private set

    // v2 sistemleri
    var evidence = 0; private set                    // kanıt sayacı
    var showLawsuitOffer = false; private set        // "dava aç?" özel kartı
    val relationships = mutableMapOf<String, Int>()  // karakter ilişki hafızası
    val legalTally = mutableMapOf<String, Int>()     // kategori bazlı "ezen karar" sayısı
    private var solidarityShield = 0
    private var lawsuitFloor = 0
    private var unionTriggered = false

    private val projects = listOf("Atlas", "Phoenix", "Nova", "Titan", "Orion", "Vega", "Zenith", "Delta-9")
    private val clients = listOf("GlobalCorp", "Meridian AŞ", "NorthBridge", "Vertex Ltd", "OmniTrade", "BlueRock")

    private val victimCats = setOf("SAG", "IZO", "ITB", "IS", "YOU")

    private val overlay: Map<String, LocCard>
    @Serializable data class LocCard(val t: String, val l: String, val r: String)
    @Serializable data class LocFile(val cards: Map<String, LocCard> = emptyMap())

    init {
        val json = Json { ignoreUnknownKeys = true }
        val files = listOf("cards_core.json", "cards_ext.json", "cards_ext2.json")
        val cards = mutableListOf<Card>()
        for (f in files) {
            val raw = context.assets.open(f).bufferedReader().use { it.readText() }
            cards += json.decodeFromString<CardFile>(raw).cards
        }
        allCards = cards
        followupIds = cards.flatMap { listOfNotNull(it.l.next, it.r.next) }.toSet()
        overlay = if (lang !in listOf("en", "tr")) {
            try {
                val raw = context.assets.open("loc_$lang.json").bufferedReader().use { it.readText() }
                json.decodeFromString<LocFile>(raw).cards
            } catch (_: Exception) { emptyMap() }
        } else emptyMap()
        drawNext()
    }

    private val recent = ArrayDeque<String>()
    private fun mainPool() = allCards.filter {
        it.minB == null && it.id !in followupIds && (it.minD ?: 0) <= day
    }
    private fun pressurePool() = allCards.filter { (it.minB ?: 999) <= meters.b }

    private fun substitute(s: String): String = s
        .replace("{P}", projects[rng.nextInt(projects.size)])
        .replace("{C}", clients[rng.nextInt(clients.size)])
        .replace("{X}", (rng.nextInt(5, 40) * 10).toString())

    var text = ""; private set
    var lText = ""; private set
    var rText = ""; private set

    private fun resolveTexts() {
        // Dava teklifi özel kartı — koddan gelir, JSON'da yok
        if (showLawsuitOffer) {
            text = Loc.s("lawsuit_offer_t")
            lText = Loc.s("lawsuit_offer_l")
            rText = Loc.s("lawsuit_offer_r")
            return
        }
        val c = current ?: return
        val o = overlay[c.id]
        text = substitute(o?.t ?: c.t.get(lang))
        lText = substitute(o?.l ?: c.l.t.get(lang))
        rText = substitute(o?.r ?: c.r.t.get(lang))
    }

    private fun drawNext() {
        pickNext()
        resolveTexts()
    }

    private fun pickNext() {
        // Sendika olayı: ekip güçlü + oturmuş + bir kez
        if (!unionTriggered && meters.e > 65 && day > 25) {
            val u = allCards.firstOrNull { it.id == "sendika" }
            if (u != null) { unionTriggered = true; current = u; return }
        }
        // 1) zincir kuyruğu öncelikli (%60)
        if (queue.isNotEmpty() && rng.nextFloat() < 0.6f) {
            val id = queue.removeFirst()
            current = allCards.firstOrNull { it.id == id } ?: return pickNext()
            return
        }
        // 2) baskı yüksekse "sana mobbing" destesi (%35)
        if (meters.b >= 75 && rng.nextFloat() < 0.35f) {
            val p = pressurePool()
            if (p.isNotEmpty()) { current = p[rng.nextInt(p.size)]; return }
        }
        // 3) güne uygun havuzdan çek (son 12 kart tekrarlanmaz)
        val pool = mainPool().filter { it.id !in recent }
        val pick = if (pool.isNotEmpty()) pool[rng.nextInt(pool.size)]
                   else mainPool()[rng.nextInt(mainPool().size)]
        recent.addLast(pick.id)
        if (recent.size > 12) recent.removeFirst()
        current = pick
    }

    var ended: Ending? = null; private set

    /** Seçim uygula. true dönerse oyun devam ediyor; false ise ended/lawsuitOffer dolu. */
    fun choose(left: Boolean): Boolean {
        // Dava teklifi ekranındaki seçim
        if (showLawsuitOffer) {
            showLawsuitOffer = false
            if (left) { ended = Ending.LAWSUIT; return false }        // dava aç → adalet (kanıt sağlam)
            lawsuitFloor = evidence + 8; drawNext(); return true      // devam: daha çok kanıt topla
        }
        val c = current ?: return false
        val ch = if (left) c.l else c.r
        val fx = ch.fx.map { it + if (it != 0) rng.nextInt(-2, 3) else 0 }

        meters.b = clamp(meters.b + fx[0]); meters.v = clamp(meters.v + fx[1])
        meters.e = clamp(meters.e + fx[2]); meters.k = clamp(meters.k + fx[3])

        // 2. KANIT: mağduriyet kartında direniş → belge. AĞIRLIKLI: sıradan direniş az,
        //    nitelikli/ağır ihlal (sağlık, sistematik) çok değer. Firma büyük — küçük kanıt yetmez.
        //    RİSKLİ: firma seni izliyor. Kanıt biriktikçe iz artar, yakalanabilirsin.
        if (c.cat in victimCats && fx[0] < 0) {
            val riskP = 0.015 + evidence * 0.0015
            if (rng.nextDouble() < riskP) {
                meters.b = clamp(meters.b + 15)        // gözetim/şüphe → baskı sıçraması
                if (evidence >= 30 && rng.nextDouble() < 0.3) {
                    evidence = 0
                    ended = Ending.CAUGHT              // yakalandın → belgeler imha, mahvoldun
                    return false
                }
                evidence /= 2                          // belgelerinin bir kısmı ele geçti
            } else {
                var w = evidenceWeight(c.cat)
                if (meters.e > 60) w += 1              // yanında tanıklar var
                if (fx[0] <= -4) w += 1                // güçlü, net bir karşı koyuş
                evidence += w
            }
        }

        // 3. İLİŞKİ HAFIZASI
        if (fx[2] > 0) relationships[c.ch] = (relationships[c.ch] ?: 0) + 1
        else if (fx[2] < 0) relationships[c.ch] = (relationships[c.ch] ?: 0) - 1

        // 4. GERÇEK KARŞILIK: "ezen" karar → kategori sayacı
        if (fx[0] > 0 && c.cat != "YOU") legalTally[c.cat] = (legalTally[c.cat] ?: 0) + 1

        // 5. SENDİKA kalkanı → güçlü toplu kanıt
        if (c.id == "sendika" && fx[0] < 0) { solidarityShield = 5; evidence += 4 }

        ch.next?.let { if (it !in queue) queue.addLast(it) }
        day++

        // 1. SAĞLIK: gün geçtikçe beden daha az dayanır
        val worn = day > 30
        var hd = 0
        if (meters.b > (if (worn) 55 else 65)) hd -= 3 else if (meters.b > 45) hd -= 1
        if (meters.v < 25) hd -= 2
        if (day > 45) hd -= 1                          // kronik yorgunluk
        if (fx[0] < 0 && fx[2] > 0) hd += 1
        meters.h = clamp(meters.h + hd)

        // Baskı creep gün geçtikçe hızlanır (sendika kalkanı varsa durur)
        if (solidarityShield > 0) solidarityShield--
        else {
            val creep = when { day < 7 -> 0; day < 18 -> 1; day < 35 -> 2; else -> 3 }
            meters.b = clamp(meters.b + creep)
        }

        ended = checkEnd()
        if (ended != null) return false

        // 2b. Dava teklifi: büyük firma — eşik yüksek ve gün geçtikçe daha da yükselir.
        //     Ancak yeterince AĞIR ve NİTELİKLİ kanıt biriktirdiysen teklif gelir → kazanacağın davadır.
        val threshold = maxOf(lawsuitFloor, 45 + day / 3)
        if (evidence >= threshold) {
            showLawsuitOffer = true
            resolveTexts()
            return true
        }
        drawNext()
        return true
    }

    private fun clamp(x: Int) = max(0, min(100, x))

    /** Kanıtın ağırlığı — hangi ihlal türü mahkemede ne kadar ağır basar */
    private fun evidenceWeight(cat: String): Int = when (cat) {
        "SAG" -> 3                    // sağlık ihlali / rapor reddi — en ağır belge
        "IS", "IZO", "ITB" -> 2
        else -> 1                     // YOU — tek başına zayıf
    }

    fun revive() {
        when (ended) {
            Ending.B0 -> meters.b = 30
            Ending.B100 -> meters.b = 70
            Ending.V0 -> meters.v = 30
            Ending.V100 -> meters.v = 70
            Ending.E0 -> meters.e = 30
            Ending.E100 -> meters.e = 70
            Ending.K0 -> meters.k = 30
            Ending.K100 -> meters.k = 70
            Ending.H0 -> meters.h = 40
            Ending.CAUGHT -> meters.b = 45   // rüşvetle kovulmaktan döndün, kanıt yok
            Ending.LAWSUIT -> return  // adalet sonu geri alınmaz
            null -> return
        }
        ended = null
        drawNext()
    }

    private fun checkEnd(): Ending? = when {
        meters.h <= 0 -> Ending.H0
        meters.b <= 0 -> Ending.B0;   meters.b >= 100 -> Ending.B100
        meters.v <= 0 -> Ending.V0;   meters.v >= 100 -> Ending.V100
        meters.e <= 0 -> Ending.E0;   meters.e >= 100 -> Ending.E100
        meters.k <= 0 -> Ending.K0;   meters.k >= 100 -> Ending.K100
        else -> null
    }

    // ----- Oyun sonu özetleri (UI için) -----
    val lovedCount: Int get() = relationships.values.count { it >= 3 }
    val hatedCount: Int get() = relationships.values.count { it <= -3 }

    /** Gerçek karşılık satırları */
    fun legalConsequences(): List<String> {
        val out = mutableListOf<String>()
        for (cat in listOf("SAG", "IZO", "ITB", "IS")) {
            val n = legalTally[cat] ?: 0
            if (n > 0) out.add(String.format(Loc.s("legal_${cat.lowercase()}"), n))
        }
        return out
    }
}

File: ios/Mobbing/Sources/GameEngine.swift
import Foundation

// ---------------------------------------------------------------------------
// Veri modeli — cards_*.json ile aynı şema (DEĞİŞMEDİ, geriye uyumlu)
// ---------------------------------------------------------------------------
struct LocText: Codable {
    let en: String
    let tr: String?
    func get(_ lang: String) -> String { lang == "tr" ? (tr ?? en) : en }
}

struct Choice: Codable {
    let t: LocText
    let fx: [Int]
    let next: String?
}

struct Card: Codable, Identifiable {
    let id: String
    let cat: String
    let ch: String
    let minB: Int?
    let minD: Int?
    let t: LocText
    let l: Choice
    let r: Choice
}

struct CardFile: Codable { let cards: [Card] }

// 5 gösterge: baskı, vicdan, ekip, kariyer, SAĞLIK (yeni)
struct Meters {
    var b = 50, v = 50, e = 50, k = 50, h = 100
}

enum Ending: String {
    case b0, b100, v0, v100, e0, e100, k0, k100
    case h0        // YENİ: sağlık çöküşü (tükenmişlik)
    case lawsuit   // YENİ: dava zaferi (adalet)
}

// ---------------------------------------------------------------------------
// Motor
// ---------------------------------------------------------------------------
final class GameEngine: ObservableObject {
    @Published var meters = Meters()
    @Published var day = 1
    @Published var current: Card?
    @Published var ended: Ending?

    // v2 sistemleri
    @Published var evidence = 0                 // kanıt sayacı
    @Published var showLawsuitOffer = false     // "dava aç?" özel kartı gösteriliyor mu
    private(set) var relationships: [String: Int] = [:]   // karakter ilişki hafızası
    private(set) var legalTally: [String: Int] = [:]      // kategori bazlı "ezen karar" sayısı
    private var solidarityShield = 0            // sendika kalkanı (tur sayısı)
    private var lawsuitFloor = 0                 // dava teklifi eşik tabanı (reddedilince yükselir)
    private var unionTriggered = false          // sendika olayı bir kez

    private let lang: String
    private var allCards: [Card] = []
    private var followupIds: Set<String> = []
    private var queue: [String] = []

    private let projects = ["Atlas", "Phoenix", "Nova", "Titan", "Orion", "Vega", "Zenith", "Delta-9"]
    private let clients = ["GlobalCorp", "Meridian AŞ", "NorthBridge", "Vertex Ltd", "OmniTrade", "BlueRock"]

    // Çeviri katmanı
    struct LocCard: Codable { let t: String; let l: String; let r: String }
    struct LocFile: Codable { let cards: [String: LocCard] }
    private var overlay: [String: LocCard] = [:]

    // Mağduriyet kategorileri (kanıt/gerçek karşılık için)
    private let victimCats: Set<String> = ["SAG", "IZO", "ITB", "IS", "YOU"]

    init(lang: String) {
        self.lang = lang
        for name in ["cards_core", "cards_ext", "cards_ext2"] {
            if let url = Bundle.main.url(forResource: name, withExtension: "json"),
               let data = try? Data(contentsOf: url),
               let file = try? JSONDecoder().decode(CardFile.self, from: data) {
                allCards += file.cards
            }
        }
        if !["en", "tr"].contains(lang),
           let url = Bundle.main.url(forResource: "loc_\(lang)", withExtension: "json"),
           let data = try? Data(contentsOf: url),
           let file = try? JSONDecoder().decode(LocFile.self, from: data) {
            overlay = file.cards
        }
        followupIds = Set(allCards.flatMap { [$0.l.next, $0.r.next].compactMap { $0 } })
        drawNext()
    }

    // ----- Kart havuzu -----
    private var recent: [String] = []
    private var mainPool: [Card] { allCards.filter {
        $0.minB == nil && !followupIds.contains($0.id) && ($0.minD ?? 0) <= day
    } }
    private var pressurePool: [Card] { allCards.filter { ($0.minB ?? 999) <= meters.b } }

    private func substitute(_ s: String) -> String {
        var out = s
        if out.contains("{P}") { out = out.replacingOccurrences(of: "{P}", with: projects.randomElement()!) }
        if out.contains("{C}") { out = out.replacingOccurrences(of: "{C}", with: clients.randomElement()!) }
        if out.contains("{X}") { out = out.replacingOccurrences(of: "{X}", with: String(Int.random(in: 5...40) * 10)) }
        return out
    }

    @Published var text = ""
    @Published var lText = ""
    @Published var rText = ""

    private func resolveTexts() {
        // Dava teklifi özel kartı — koddan gelir, JSON'da yok
        if showLawsuitOffer {
            text = L("lawsuit_offer_t")
            lText = L("lawsuit_offer_l")
            rText = L("lawsuit_offer_r")
            return
        }
        guard let c = current else { return }
        let o = overlay[c.id]
        text = substitute(o?.t ?? c.t.get(lang))
        lText = substitute(o?.l ?? c.l.t.get(lang))
        rText = substitute(o?.r ?? c.r.t.get(lang))
    }

    private func drawNext() {
        pickNext()
        resolveTexts()
    }

    private func pickNext() {
        // Sendika olayı: ekip güçlü + oturmuş + bir kez
        if !unionTriggered && meters.e > 65 && day > 25,
           let u = allCards.first(where: { $0.id == "sendika" }) {
            unionTriggered = true; current = u; return
        }
        if !queue.isEmpty && Double.random(in: 0...1) < 0.6 {
            let id = queue.removeFirst()
            if let c = allCards.first(where: { $0.id == id }) { current = c; return }
        }
        if meters.b >= 75 && Double.random(in: 0...1) < 0.35, let c = pressurePool.randomElement() {
            current = c; return
        }
        let pool = mainPool.filter { !recent.contains($0.id) }
        let pick = pool.randomElement() ?? mainPool.randomElement()!
        recent.append(pick.id)
        if recent.count > 12 { recent.removeFirst() }
        current = pick
    }

    // ----- Seçim -----
    func choose(left: Bool) {
        // Dava teklifi ekranındaki seçim
        if showLawsuitOffer {
            showLawsuitOffer = false
            if left { ended = .lawsuit }               // dava aç → adalet sonu (kanıt yeterince güçlü)
            else { lawsuitFloor = evidence + 8; drawNext() }  // devam: dava için daha çok kanıt topla
            return
        }
        guard let c = current else { return }
        let ch = left ? c.l : c.r
        let fx = ch.fx.map { $0 == 0 ? 0 : $0 + Int.random(in: -2...2) }

        meters.b = clamp(meters.b + fx[0]); meters.v = clamp(meters.v + fx[1])
        meters.e = clamp(meters.e + fx[2]); meters.k = clamp(meters.k + fx[3])

        // 2. KANIT: mağduriyet kartında direniş → belge. AĞIRLIKLI: sıradan direniş az,
        //    nitelikli/ağır ihlal (sağlık, sistematik) çok değer. Firma büyük — küçük kanıt yetmez.
        if victimCats.contains(c.cat) && fx[0] < 0 {
            var w = evidenceWeight(c.cat)
            if meters.e > 60 { w += 1 }                // yanında tanıklar var
            if fx[0] <= -4 { w += 1 }                  // güçlü, net bir karşı koyuş
            evidence += w
        }

        // 3. İLİŞKİ HAFIZASI: karaktere yönelik ekip etkisi
        if fx[2] > 0 { relationships[c.ch, default: 0] += 1 }
        else if fx[2] < 0 { relationships[c.ch, default: 0] -= 1 }

        // 4. GERÇEK KARŞILIK: "ezen" (baskı artıran) karar → kategori sayacı
        if fx[0] > 0 && c.cat != "YOU" { legalTally[c.cat, default: 0] += 1 }

        // 5. SENDİKA kalkanı: sendika kartında birlik (baskı düşür) seçilirse → güçlü toplu kanıt
        if c.id == "sendika" && fx[0] < 0 { solidarityShield = 5; evidence += 4 }

        if let nx = ch.next, !queue.contains(nx) { queue.append(nx) }
        day += 1

        // 1. SAĞLIK: baskı ve vicdana bağlı otomatik erime — gün geçtikçe beden daha az dayanır
        let worn = day > 30                            // uzun süre = yıpranmış eşik
        var hd = 0
        if meters.b > (worn ? 55 : 65) { hd -= 3 } else if meters.b > 45 { hd -= 1 }
        if meters.v < 25 { hd -= 2 }
        if day > 45 { hd -= 1 }                        // kronik yorgunluk
        if fx[0] < 0 && fx[2] > 0 { hd += 1 }          // insani karar = nefes
        meters.h = clamp(meters.h + hd)

        // Sistem yorulmaz: baskı creep gün geçtikçe hızlanır (sendika kalkanı varsa durur)
        if solidarityShield > 0 { solidarityShield -= 1 }
        else {
            let creep = day < 7 ? 0 : (day < 18 ? 1 : (day < 35 ? 2 : 3))
            meters.b = clamp(meters.b + creep)
        }

        // 2b. Dava teklifi: eşik gün geçtikçe yükselir (firma avukatları güçlenir).
        //     Ancak yeterince SAĞLAM kanıt biriktirdiysen teklif gelir → kazanacağın davadır.
        ended = checkEnd()
        if ended != nil { return }
        // Büyük firma: dava eşiği yüksek ve gün geçtikçe daha da yükselir (firma avukatları güçlenir).
        // Ancak yeterince AĞIR ve NİTELİKLİ kanıt biriktirdiysen teklif gelir — kazanacağın davadır.
        let threshold = max(lawsuitFloor, 55 + day / 3)
        if evidence >= threshold {
            showLawsuitOffer = true
            resolveTexts()
            return
        }
        drawNext()
    }

    private func clamp(_ x: Int) -> Int { max(0, min(100, x)) }

    /// Kanıtın ağırlığı — hangi ihlal türü mahkemede ne kadar ağır basar
    private func evidenceWeight(_ cat: String) -> Int {
        switch cat {
        case "SAG": return 3          // sağlık ihlali / rapor reddi — en ağır belge
        case "IS", "IZO", "ITB": return 2
        default: return 1             // YOU (sana yönelik) — tek başına zayıf
        }
    }

    func forceEnd(_ e: Ending) { ended = e }

    /// Rüşvet: seni bitiren göstergeyi güvenli bölgeye çeker
    func revive() {
        guard let end = ended else { return }
        switch end {
        case .b0: meters.b = 30;  case .b100: meters.b = 70
        case .v0: meters.v = 30;  case .v100: meters.v = 70
        case .e0: meters.e = 30;  case .e100: meters.e = 70
        case .k0: meters.k = 30;  case .k100: meters.k = 70
        case .h0: meters.h = 40
        case .lawsuit: return   // adalet sonu geri alınmaz
        }
        ended = nil
        drawNext()
    }

    private func checkEnd() -> Ending? {
        if meters.h <= 0 { return .h0 }
        if meters.b <= 0 { return .b0 };   if meters.b >= 100 { return .b100 }
        if meters.v <= 0 { return .v0 };   if meters.v >= 100 { return .v100 }
        if meters.e <= 0 { return .e0 };   if meters.e >= 100 { return .e100 }
        if meters.k <= 0 { return .k0 };   if meters.k >= 100 { return .k100 }
        return nil
    }

    // ----- Oyun sonu özetleri (UI için) -----
    /// Sevilen/nefret edilen karakter sayısı
    var lovedCount: Int { relationships.values.filter { $0 >= 3 }.count }
    var hatedCount: Int { relationships.values.filter { $0 <= -3 }.count }

    /// Gerçek karşılık satırları — verilen "ezen" kararların iş hukuku karşılığı
    func legalConsequences() -> [String] {
        var out: [String] = []
        let order = ["SAG", "IZO", "ITB", "IS"]
        for cat in order {
            if let n = legalTally[cat], n > 0 {
                out.append(String(format: L("legal_\(cat.lowercased())"), n))
            }
        }
        return out
    }
}
